go/parquet/metadata/app_version.go

```diff
@@ -0,0 +1,172 @@
+	parquetCPPFixedStatsVersion = NewAppVersionExplicit("parquet-cpp", 1, 3, 0)
+	parquetMRFixedStatsVersion  = NewAppVersionExplicit("parquet-mr", 1, 10, 0)
+)
+
+// AppVersion represents a specific application version either read from
+// or written to a parquet file.
+type AppVersion struct {
+	App     string
+	Build   string
+	Version struct {
+		Major      int
+		Minor      int
+		Patch      int
+		Unknown    string
+		PreRelease string
+		BuildInfo  string
+	}
+}
+
+// NewAppVersionExplicit is a convenience function to construct a specific
+// application version from the given app string and version
+func NewAppVersionExplicit(app string, major, minor, patch int) *AppVersion {
+	v := &AppVersion{App: app}
+	v.Version.Major = major
+	v.Version.Minor = minor
+	v.Version.Patch = patch
+	return v
+}
+
+// NewAppVersion parses a "created by" string such as "parquet-go 1.0.0".
+//
+// It also supports handling pre-releases and build info such as
+// 	parquet-cpp version 1.5.0ab-xyz5.5.0+cd (build abcd)
+func NewAppVersion(createdby string) *AppVersion {
+	v := &AppVersion{}
+
+	var ver []string
+
+	m := applicationRx.FindStringSubmatch(strings.ToLower(createdby))
+	if len(m) >= 4 {
+		v.App = m[1]
+		v.Build = m[4]
+		ver = versionRx.FindStringSubmatch(m[3])
+	} else {
+		v.App = "unknown"
+	}
+
+	if len(ver) >= 7 {
+		v.Version.Major, _ = strconv.Atoi(ver[1])
+		v.Version.Minor, _ = strconv.Atoi(ver[2])
+		v.Version.Patch, _ = strconv.Atoi(ver[3])
+		v.Version.Unknown = ver[4]
+		v.Version.PreRelease = ver[5]
+		v.Version.BuildInfo = ver[6]
+	}
+	return v
+}
+
+// LessThan compares the app versions and returns true if this version
+// is "less than" the passed version.
+//
+// If the apps don't match, this always returns false. Otherwise it compares
+// the major versions first, then the minor versions, and finally the patch
+// versions.
+//
+// Pre-release and build info are not considered.
+func (v AppVersion) LessThan(other *AppVersion) bool {
+	switch {
+	case v.App != other.App:
+		return false
+	case v.Version.Major < other.Version.Major:
+		return true
+	case v.Version.Major > other.Version.Major:
+		return false
+	case v.Version.Minor < other.Version.Minor:
+		return true
+	case v.Version.Minor > other.Version.Minor:
+		return false
+	}
+
+	return v.Version.Patch < other.Version.Patch
+}
+
+// Equal only compares the Application and major/minor/patch versions.
+//
+// Pre-release and build info are not considered.
+func (v AppVersion) Equal(other *AppVersion) bool {
+	return v.App == other.App &&
+		v.Version.Major == other.Version.Major &&
+		v.Version.Minor == other.Version.Minor &&
+		v.Version.Patch == other.Version.Patch
+}
+
+// HasCorrectStatistics checks whether or not the statistics are valid to be used
+// based on the primitive type and the version since previous versions had issues with
+// properly computing stats.
+//
+// Reference: parquet-cpp/src/parquet/metadata.cc
+//
+// PARQUET-686 has more discussion on statistics
```

Comment:
We haven't done this elsewhere but it isn't the only statistics bug from he C++ perspective.  Decimal logical types comparisons for C++ where broken until recently and null statistics for list types are still broken.

Reply:
added check for parquet-cpp-arrow version 4.0.0 to check decimal stats etc.